llvmPackages stdenv broken on master and unstable, can't find C++ headers?

## Issue description

The clang stdenv (e.g., `llvmPackages_4.stdenv`) on recent unstable channel no longer seems able to find the C++ headers when used as the stdenv for a build.



### Steps to reproduce

```
$ NIX_PATH=unstable=https://d3g5gsiof5omrk.cloudfront.net/nixos/unstable/nixos-17.09pre112691.abdb58e407/nixexprs.tar.xz nix-build -E "with import <unstable> {}; lzip.override { stdenv = llvmPackages_4.stdenv; }"
```

yields

```
...
build flags: SHELL=/nix/store/8v2glga3sg78wj5fmxqvf63vdxrjgy6h-bash-4.4-p12/bin/bash
c++ -DNDEBUG -O3 -c -o arg_parser.o arg_parser.cc
arg_parser.cc:20:10: fatal error: 'cstring' file not found
#include <cstring>
         ^~~~~~~~~
1 error generated.
make: *** [Makefile:50: arg_parser.o] Error 1
```

Curiously the libcxxStdenv does work:

```
NIX_PATH=unstable=https://d3g5gsiof5omrk.cloudfront.net/nixos/unstable/nixos-17.09pre112691.abdb58e407/nixexprs.tar.xz nix-build -E "with import <unstable> {}; lzip.override { stdenv = llvmPackages_4.libcxxStdenv; }"
```

and just for sanity, it does work on 17.03 and usually works otherwise O:)

```
NIX_PATH=nixpkgs=https://nixos.org/channels/nixos-17.03/nixexprs.tar.xz nix-build -E "with import <nixpkgs> {}; lzip.override { inherit (llvmPackages_4) stdenv; }"
```

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
* Nix version: (run `nix-env --version`)
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
* Sandboxing enabled: (run `grep build-use-sandbox /etc/nix/nix.conf`)

--- a/pkgs/development/compilers/gcc/libstdc++-hook.sh
+++ b/pkgs/development/compilers/gcc/libstdc++-hook.sh
@@ -1,0 +1,1 @@
+export NIX_CXXSTDLIB_COMPILE+="-isystem $(echo -n @gcc@/include/c++/*) -isystem $(echo -n @gcc@/include/c++/*)/$(@gcc@/bin/gcc -dumpmachine)"

--- a/pkgs/development/compilers/gcc/libstdc++-hook.sh
+++ b/pkgs/development/compilers/gcc/libstdc++-hook.sh
@@ -1,1 +1,2 @@
 export NIX_CXXSTDLIB_COMPILE+="-isystem $(echo -n @gcc@/include/c++/*) -isystem $(echo -n @gcc@/include/c++/*)/$(@gcc@/bin/gcc -dumpmachine)"
+export NIX_CXXSTDLIB_LINK=" -stdlib=libstdc++"

--- a/pkgs/development/compilers/llvm/4/default.nix
+++ b/pkgs/development/compilers/llvm/4/default.nix
@@ -1,4 +1,8 @@
-{ lowPrio, newScope, stdenv, cmake, libxml2, python2, isl, fetchurl, overrideCC, wrapCC, darwin, ccWrapperFun }:
+{ lowPrio, newScope, stdenv, targetPlatform, cmake, libstdcxxHook
+, libxml2, python2, isl, fetchurl, overrideCC, wrapCC, ccWrapperFun
+, darwin
+}:
+
 let
   callPackage = newScope (self // { inherit stdenv cmake libxml2 python2 isl release_version version fetch; });
 
@@ -46,7 +50,12 @@ let
       extraPackages = [ self.libcxx self.libcxxabi ];
     };
 
-    stdenv = overrideCC stdenv self.clang;
+    stdenv = stdenv.override (drv: {
+      allowedRequisites = null;
+      cc = self.clang;
+      # Use the gcc libstdc++ when targeting linux.
+      extraBuildInputs = if stdenv.cc.isGNU then [ libstdcxxHook ] else drv.extraBuildInputs;
+    });
 
     libcxxStdenv = overrideCC stdenv self.libcxxClang;
 
@@ -58,4 +67,5 @@ let
 
     libcxxabi = callPackage ./libc++abi.nix {};
   };
+
 in self

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -5354,6 +5354,10 @@ with pkgs;
 
   gccApple = throw "gccApple is no longer supported";
 
+  libstdcxxHook = makeSetupHook
+    { substitutions = { gcc = gcc-unwrapped; }; }
+    ../development/compilers/gcc/libstdc++-hook.sh;
+
   # Can't just overrideCC, because then the stdenv-cross mkDerivation will be
   # thrown away. TODO: find a better solution for this.
   crossLibcStdenv = buildPackages.makeStdenvCross {

